NXP: add SPSDK as runner and support i.MX95 firmware build and flashing

yangbolu1991 opened the pull request:
This PR is to make it easier to build and flash NXP i.MX platforms having firmware images dependency with NXP SPSDK (https://spsdk.readthedocs.io).
For i.MX95, an bootable image depends on many firmware images besides m7 zephyr.bin, like ddr firmware, system manager image, oei images, ahab container image and so on.

This PR is to support west build  to generate a bootable flash.bin.
And west flash to program flash.bin to SD boot device through USB.

Thanks.

yangbolu1991 committed:
boards: nxp: imx95_evk: support generating/flashing bootable firmware

Supported NXP Secure Provisioning SDK (SPSDK) generating
and flashing bootable firmware for imx95_evk_mimx9596_m7/a55.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

yangbolu1991 committed:
boards: common: add spsdk flasher support

Added spsdk flasher support.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

hakehuang commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,11 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+
+if (CONFIG_BOARD_IMX95_EVK_MIMX9596_M7 AND CONFIG_BOARD_NXP_SPSDK_IMAGE)
+  set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+    COMMAND cp -r ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/mimx9596/imx95-19x19-lpddr5-evk ${CMAKE_BINARY_DIR}/zephyr/
```
configure_file(<input> <output> COPYONLY)

yangbolu1991 replied:
Thanks.
I will point to source path without copying.

hakehuang commented on boards/nxp/common/Kconfig:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_NXP_SPSDK_IMAGE
+	bool "Generate bootable firmware image with SPSDK"
```
spsdk comes with version, so it is better to add a version check

yangbolu1991 replied:
Let me limit it to verified versions in scripts/requirements-extras.txt

hakehuang commented on boards/nxp/imx95_evk/board.cmake:
```diff
@@ -3,3 +3,12 @@
 if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
   board_set_rimage_target(imx95)
 endif()
+
```
we need add document on how to use spsdk on imx95.

yangbolu1991 replied:
No problem. Thanks.

nordicjm commented on boards/nxp/common/Kconfig:
```diff
@@ -0,0 +1,9 @@
+
+config BOARD_NXP_SPSDK_IMAGE
+	bool "Generate bootable firmware image with SPSDK"
+	help
```
`depends on CONFIG_BOARD_IMX95_EVK_MIMX9596_M7`

yangbolu1991 replied:
Okay.

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,11 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+if (CONFIG_BOARD_IMX95_EVK_MIMX9596_M7 AND CONFIG_BOARD_NXP_SPSDK_IMAGE)
+  set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+    COMMAND cp -r ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/mimx9596/imx95-19x19-lpddr5-evk ${CMAKE_BINARY_DIR}/zephyr/
+    COMMAND nxpimage ahab export -c ${PROJECT_SOURCE_DIR}/boards/nxp/imx95_evk/imx95_evk_mimx9596_m7_ahab.yaml
```
this is not going to work on windows since cp is a *nix command

yangbolu1991 replied:
Will point to source path without copying.

nordicjm commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_m7_ahab.yaml:
```diff
@@ -0,0 +1,356 @@
+  # ====================================================================================================================
+
+  #  =========================== [Example of possible configuration: #0 , erase if not used] ============================
+    # -----------------------------------===== Binary AHAB container [Required] =====-----------------------------------
```
this file cannot be here, it will be treated as a twister file

yangbolu1991 replied:
WIll remove it and generate during build.

yangbolu1991 commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,11 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
 endif()
+
+
+if (CONFIG_BOARD_IMX95_EVK_MIMX9596_M7 AND CONFIG_BOARD_NXP_SPSDK_IMAGE)
```
@nordicjm 
I keep CONFIG_BOARD_IMX95_EVK_MIMX9596_M7 here since it's board specific.
Other boards will add support too.

yangbolu1991 commented:
Updated to v2. Changes include,
- Dropped images copy.
- Dropped yml file and generated during build instead.
- Improved Kconfig.
- Added doc.
- Fixed CI issues.

Thank you very much @hakehuang  and @nordicjm

nordicjm commented:
west command CI is failing

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,33 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
   )
 endif()
+
+if (CONFIG_BOARD_NXP_SPSDK_IMAGE)
+  if (CONFIG_BOARD_IMX95_EVK_MIMX9596_M7)
```
```suggestion
if(CONFIG_BOARD_NXP_SPSDK_IMAGE AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7)
```

yangbolu1991 replied:
Ok, thanks.

nordicjm replied:
Not addressed

nordicjm commented on boards/nxp/common/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_NXP_SPSDK_IMAGE
+	bool "Generate bootable firmware image with SPSDK"
```
Is this something this SoC will use, i.e. on any board? If so, should it go at soc level rather than board?

nordicjm replied:
Sure

hakehuang commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -233,13 +233,50 @@ To program M7, an i.MX container image ``flash.bin`` must be made, which contain
 image and so on.
 
-The steps making flash.bin and programming should refer to ``Getting Started with
+Two methods to build and program ``flash.bin``. Either can be choosed.
```
```suggestion
Two methods to build and program ``flash.bin``. Either can be choosed.
board need set to usb serial download mode SW7 1001, and using the first serial port enumerated.
```

yangbolu1991 replied:
I will mention the serial port information.
usb serial download mode setting had been described in method2.
For method1, doc just guides to refer to MCUX SDK doc for users who may be familar with SDK. They may be not through USB serial download.

Thanks.

hakehuang commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -233,13 +233,50 @@ To program M7, an i.MX container image ``flash.bin`` must be made, which contain
+
+   # load and run without programming
+   $ west flash --bootdevice=spl
+
```
```suggestion
For next flash, you need reset the board by enter 'reset' in the 4th enumerated serial port.
```

yangbolu1991 replied:
okay.

hakehuang commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -233,13 +233,50 @@ To program M7, an i.MX container image ``flash.bin`` must be made, which contain
+   $ west flash
+
+   # program to emmc card
+   $ west flash --bootdevice=emmc
```
```suggestion
you need switch the board to SD/MMC mode, SW7 1010, and power cycle board to boot with it.
```

yangbolu1991 commented:
> west command CI is failing

Will resolve that. Thanks.

yangbolu1991 commented:
Updated to v3. Changes include,
- Improved cmake file per @nordicjm  suggestion.
- Improved doc per @hakehuang  suggestion.
- Fixed CI west command issues.

Thank you very much.

pdgendt commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,31 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+      - container:
+          srk_set: none
+          images:
+            - lpddr_imem:     ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/lpddr5_imem_v202311.bin
+              lpddr_imem_qb:  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/lpddr5_imem_qb_v202311.bin
+              lpddr_dmem:     ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/lpddr5_dmem_v202311.bin
+              lpddr_dmem_qb:  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/lpddr5_dmem_qb_v202311.bin
+              oei_ddr:        ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/oei-m33-ddr.bin
+            - oei_tcm:        ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/oei-m33-tcm.bin
+            - system_manager: ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/m33_image-mx95alt.bin
```
I'd suggest adding the following to verify the blobs's integrity.
```cmake
zephyr_blobs_verify(FILES
  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/lpddr5_imem_v202311.bin
  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/lpddr5_imem_qb_v202311.bin
  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/lpddr5_dmem_v202311.bin
  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/lpddr5_dmem_qb_v202311.bin
  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/oei-m33-ddr.bin
  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/oei-m33-tcm.bin
  ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/m33_image-mx95alt.bin
  REQUIRED)
```

yangbolu1991 replied:
The verification is good.
But for now, west blobs hasn't been supported. It may take very long time to do that.

So, CONFIG_BOARD_NXP_SPSDK_IMAGE is still need to open manually, and prepare images manually.
We couldn't add zephyr_blobs_verify for now.
Thanks.

JiafeiPan commented on boards/nxp/imx95_evk/board.cmake:
```diff
@@ -3,3 +3,12 @@
 if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
   board_set_rimage_target(imx95)
 endif()
+
```
Include the following items by CONFIG_BOARD_NXP_SPSDK_IMAGE
```suggestion
if (CONFIG_BOARD_NXP_SPSDK_IMAGE)
```

nordicjm replied:
no space between if and ( in cmake

yangbolu1991 replied:
Get it. Thanks.

JiafeiPan commented on boards/nxp/imx95_evk/board.cmake:
```diff
@@ -3,3 +3,12 @@
+board_runner_args(spsdk "--bootloader=${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-firmware/imx95-19x19-lpddr5-evk/imx-boot-imx95-19x19-lpddr5-evk-sd.bin-flash_a55")
+board_runner_args(spsdk "--flashbin=${CMAKE_BINARY_DIR}/zephyr/flash.bin")
+
+include(${ZEPHYR_BASE}/boards/common/spsdk.board.cmake)
```
```suggestion
include(${ZEPHYR_BASE}/boards/common/spsdk.board.cmake)
endif()
```

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,31 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
   )
 endif()
+
+if (CONFIG_BOARD_NXP_SPSDK_IMAGE AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7)
```
```suggestion
if(CONFIG_BOARD_NXP_SPSDK_IMAGE AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7)
```

yangbolu1991 replied:
Thanks, will fix.

yangbolu1991 commented:
Updated to v4. Changes include,

- More help description for CONFIG_BOARD_NXP_SPSDK_IMAGE.
- Fixed `if (` style issue in cmake file.
- Formatted python file spsdk.py.
- Default using spl mode (load and run) for flashing and updated doc.
- Added testing binary in yaml file.
- Wrappered with CONFIG_BOARD_NXP_SPSDK_IMAGE for board runner setting.

Thanks a  lot.

yangbolu1991 commented:
Updated to v4. Changes include,

- Fixed doc build.

Thanks.

JiafeiPan commented on scripts/west_commands/runners/spsdk.py:
```diff
@@ -0,0 +1,91 @@
+
+    def flash(self, **kwargs):
+        if self.bootdevice == 'spl':
+            cmd = [self.commander] + ['run'] + [f"SDPS[-t 10000]: boot -f {self.flashbin}"]
```
Actually it maybe use a uuu script to handle multiple containers, so is it possible to use a specific script according to different platform requirement?

yangbolu1991 replied:
In my opinion, west flash is to program a bootable firmware to board.
What do you mean by multiple containers? You want to flash multiple images by west flash? Or any example of the script to share
If the script could be as an argument to use, I think you can send patch to extend the support when you have specific platofrm/case to use.
Thanks.

JiafeiPan replied:
Multiple containers will be used in order to boot U-Boot and Linux, NP, let's cover this feature in future.

JiafeiPan commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -233,13 +233,54 @@ To program M7, an i.MX container image ``flash.bin`` must be made, which contain
+       ├── mx95a0-ahab-container.img
+       ├── oei-m33-ddr.bin
+       └── oei-m33-tcm.bin
+
```
I know currently we don't provide blobs to boot u-boot or Linux together with m-core u-boot simulteneously, but according to discussion in https://github.com/zephyrproject-rtos/zephyr/issues/82560, may I know whether can provide some document information  for how to do it?

yangbolu1991 replied:
For this PR, I think multicore case is another topic/work to do, as all we did now in zephyr is only A55 or M7 independently. I can't provide that as I haven't done any verification on that.

But that's important to support multicore case. The difficulty is the permission/resource allocation to M7 and A55. It's case specific.

1. Permission/resource allocation to M7 and A55 must be exclusive.
2. Resource usage on M core Zephyr, and A core Linux BSP must be exclusive.

I don't think we can provide any generic support for multicore. But that will be very good, if we can provide a specific sample to demo multicore work together with zephyr in the future.

Thanks.

JiafeiPan replied:
I understand the resource and permission allocation is case by case, but maybe can provide a very simple hello world application which use almost so less hardware resource except,memroy, UART and some Core related hardware resource, but it will cover update in Linux dts, I am okay to provides such features in future, thanks.

nordicjm commented on scripts/requirements-extras.txt:
```diff
@@ -15,6 +15,9 @@ junit2html
 # Script used to build firmware images for NXP LPC MCUs.
 lpc_checksum
 
+# used by NXP platform to generate/flash firmware images
+spsdk == 2.3.0
```
This can be moved to your hal, right @pdgendt ?

pdgendt replied:
It is used by an in-tree runner however, not sure what the "policy" is here.

pdgendt replied:
BTW, since https://github.com/zephyrproject-rtos/zephyr/pull/83190 out-of-tree runners are supported too.

yangbolu1991 replied:
Moved out of tree. Thanks.

hakehuang replied:
@pdgendt , @yangbolu1991 , @nordicjm out-of-tree runner is not support well in twister CI, it is only called in `west flash/build/debug`, however in twister CI it calls another path, because, in twister CI we support build and run in different machine, so it is hard to get the out-of -tree runner in build time, unless we always sync hal, which is what we are try to avoid. My suggestion would be move the spsdk.py back to in tree runner

nordicjm replied:
Not understanding the issue so can't comment on that, but it sounds like you are building with one version of zephyr and running tests on another machine with a different version of zephyr using the artefacts from the other machine, this is invalid, you should build and run tests with the same commits/versions of zephyr

hakehuang replied:
@nordicjm , Let me clarify, this is not about zephyr version issue, As the runner is off-tree, then we need parse the runner class, in current west code there is a call to the `do_run_common` in run_common.py, however, if we run twister with --test-only, we do not call the `do_run_common`,  instead it calls do_run_common_image directly and try to load the runner from runners.yaml, as the runnners.yaml only has in tree runner path, which is another problem as it is a absolute path in build machine. So when we do build on one machine and run it with `--test-only` in twister, out-of tree runner can not be found.

nordicjm replied:
You need the same tree on build and test machines including modules and the version of zephyr and said modules, it is not a zephyr problem nor should zephyr change because of your test setup if it does not follow this

hakehuang replied:
@nordicjm if we put the runner in tree, we will not have the trouble to update all modules, which is very time consumption for a CI system. So why not, if a simple code update can make a much more green world?

carlescufi replied:
I disagree with the notion that we should move runners to HAL repos. We want HAL repos to contain only non-Zephyr specific, RTOS-neutral code, with everything that is specifically written for Zephyr being in the main tree. That is why we have so much glue code in the `modules/` folder in the main tree. Runners are effectively glue code for Zephyr, so I think they belong in-tree for SoCs and boards that are supported in-tree by Zephyr.

mmahadevan108 replied:
@yangbolu1991 @JiafeiPan this PR still needs to address the concern from @carlescufi

JiafeiPan replied:
@carlescufi would you please revisit this PR, I believe your changes requested has been addressed, thanks.

JiafeiPan reviewed (changes requested):
Depend on hal update.

JiafeiPan commented:
> @JiafeiPan , as the blobs is there why not enable none sd/mme version? because with sd/mmc you still need to change the boot switch

We will support both download mode (no need to change boot switch) and sd/mmc mode (need to use download mode to flash and then change boot switch to boot).
And we need to update this PR after binary blobs is uploaded, so I request changes on it,  no other changes as we aligned.

yangbolu1991 commented:
I'm sorry for late update, since I was waiting the blobs uploaded to nxp.com. It's uploaded this week.
Update to v6. Changes include,
- Converted to use blobs.
- Updated copyright date.

As the out-of-tree runner, I didn't see it's supported by now, right? @pdgendt @nordicjm 
https://github.com/zephyrproject-rtos/zephyr/issues/83145
https://github.com/zephyrproject-rtos/zephyr/issues/23452

Thanks.

yangbolu1991 commented:
It looks like there were many unit tests faied, and two twister issues.
These were caused by verifying blobs.

Anyone know how to handle these CI issues, if west build requires blobs.
Thank you very much.

hakehuang commented:
> It looks like there were many unit tests faied, and two twister issues. These were caused by verifying blobs.
> 
> Anyone know how to handle these CI issues, if west build requires blobs. Thank you very much.

@yangbolu1991 you need check the raw log. e.g. west blob fetch fails.
```
-- Found assembler: /opt/toolchains/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
CMake Error at /__w/zephyr/zephyr/cmake/modules/extensions.cmake:1838 (message):
  Blob for path
  "/__w/zephyr/modules/hal/nxp/zephyr/blobs/imx-boot-firmware/imx95-19x19-lpddr5-evk-boot-firmware-0.1.bin"
  missing.  Update with: west blobs fetch
Call Stack (most recent call first):
  /__w/zephyr/zephyr/boards/nxp/imx95_evk/CMakeLists.txt:43 (zephyr_blobs_verify)
```

yangbolu1991 commented:
> > It looks like there were many unit tests faied, and two twister issues. These were caused by verifying blobs.
> > Anyone know how to handle these CI issues, if west build requires blobs. Thank you very much.
> 
> @yangbolu1991 you need check the raw log. e.g. west blob fetch fails.
> 
> ```
> -- Found assembler: /opt/toolchains/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
> CMake Error at /__w/zephyr/zephyr/cmake/modules/extensions.cmake:1838 (message):
>   Blob for path
>   "/__w/zephyr/modules/hal/nxp/zephyr/blobs/imx-boot-firmware/imx95-19x19-lpddr5-evk-boot-firmware-0.1.bin"
>   missing.  Update with: west blobs fetch
> Call Stack (most recent call first):
>   /__w/zephyr/zephyr/boards/nxp/imx95_evk/CMakeLists.txt:43 (zephyr_blobs_verify)
> ```

Not fetch fail. CI didn't fetch.
I am not sure if CI support blobs.
Thanks.

hakehuang commented:
> Not fetch fail. CI didn't fetch.

it says 
```
CMake Error at /__w/zephyr/zephyr/cmake/modules/extensions.cmake:1838 (message):
  Blob for path
  "/__w/zephyr/modules/hal/nxp/zephyr/blobs/imx-boot-firmware/imx95-19x19-lpddr5-evk-boot-firmware-0.1.bin"
```

which means you need update the zephyr/cmake/modules/extensions.cmake


> I am not sure if CI support blobs, as I saw some commit message like this,

you are right, CI scripts does not call the west update, so for now, I recommend update our CMakefile to call 'west blobs update hal_nxp'

yangbolu1991 commented:
Updated to v7. Changes,
- Default not enable SPSDK build per @hakehuang suggestion.
- Also added ENV control for SPSDK build.
- Rebased.

Thanks.

hakehuang commented:
> Thanks.

@yangbolu1991 you can remove the DMN tag if everything is ready.

JiafeiPan commented on scripts/requirements-extras.txt:
```diff
@@ -15,6 +15,9 @@ junit2html
 lpc_checksum
 
+# used by NXP platform to generate/flash firmware images
+spsdk == 2.3.0
```
Let's upversion spsdk to v2.6.0 in which a set of new parameter of nxpuuu are added, use these parameter to specify which board will be used in case of multiple boards are connected to the same host, this is important for testing.

New nxpuuu parameters provided in 2.6.o is: -up (--usbpath) or -us (--usbserial)

yangbolu1991 replied:
okay, updated the verision. But I didn't find the parameters from the command or latest doc. Could you share?

JiafeiPan replied:
If you have already installed an older version of spsdk,  you need to remove the older one and re-install the latest one:
```
python -m pip install --upgrade pip
pip uninstall spsdk
pip install spsdk
```
Then use "`spsdk --version`" to make sure 2.6.0 is install.
So that you could use "nxpuuu --help" to check that the following parameters are available:
```
  -up, --usbpath USB_PATH         Filter UUU devices by USB path
  -us, --usbserial SERIAL_NUMBER  Filter UUU devices by USB serial number
```

JiafeiPan commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,39 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+        - container:
+            srk_set: none
+            images:
+              - lpddr_imem:     ${CMAKE_BINARY_DIR}/zephyr/imx95-19x19-lpddr5-evk-boot-firmware-0.1/lpddr5_imem_v202311.bin
```
for "imx95-19x19-lpddr5-evk-boot-firmware-0.1", suggest to define a variable for firmware image name or version name, so it will be easier to update in case of version or image name update.

yangbolu1991 replied:
OKay, updated using variable.

JiafeiPan commented on boards/nxp/common/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config BOARD_NXP_SPSDK_IMAGE
+	bool "Generate bootable firmware image with SPSDK"
+	depends on BOARD_IMX95_EVK_MIMX9596_M7
```
remove dependency at here as it is 'n' by default, so that it can be enabled by board defconfig if it supports SPSDK image.

yangbolu1991 replied:
Removed.

JiafeiPan reviewed (commented):
looks fine for me except upversion spsdk version and some refines.

JiafeiPan reviewed (commented):
Hi, @pdgendt , could you please revisit this PR to check whether your comments has been addressed, thanks.

JiafeiPan commented:
> Hi, @pdgendt , could you please revisit this PR to check whether your comments has been addressed, thanks.

Just found your latest comments, thanks.

nordicjm commented on boards/nxp/common/Kconfig:
```diff
@@ -0,0 +1,10 @@
+config BOARD_NXP_SPSDK_IMAGE
+	bool "Generate bootable firmware image with SPSDK"
+	depends on BOARD_IMX95_EVK_MIMX9596_M7
+	default n
```
```suggestion
```

yangbolu1991 replied:
Removed, thanks.

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,39 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+              - v2x_dummy: true
+      ")
+    set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+      COMMAND chmod +x ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/imx95-19x19-lpddr5-evk-boot-firmware-0.1.bin
```
this doesn't work on windows

yangbolu1991 commented:
> > Thanks.
> 
> @yangbolu1991 why not add A core support in this PR as well? besides, in the board.cmake below need read from env as well
> 
> ```
> if(CONFIG_BOARD_NXP_SPSDK_IMAGE)
> ```
> 
> or
> 
> you can remove this if, as it is just for flash configurations, which is required anyway, when flash by twister.

okay

yangbolu1991 commented:
Updated to v8. Changes,
- Moved spsdk runner out of tree.
- Required spsdk 2.6.0 and allow to use in linux for now.
- Improvement/fixes in cmake/kconfig files per @hakehuang @JiafeiPan @nordicjm  comments.

Thanks a lot.

hakehuang commented:
there is one issue found here

```
/home/jenkins/agent/workspace/zephyr_test_pipe/zephyr/twister-out/imx95_evk_mimx9596_m7/zephyr/tests/misc/test_build/buildsystem.kconfig.utf8_in_values/zephyr/A_kconfig_value_with_a_utf8_char_in_it_Be_.elf for board: imx95_evk Welcome to NXP imx95-19x19-lpddr5-evk-boot-firmware-0.1.bin WARNING: EULA has been auto-accepted; this implies that you agree with it. Unpacking file ............... done SPSDKError: SPSDK: Configuration validation failed: data.containers[1] must be valid exactly by one definition (0 matches found) Reason of fail for oneOf rule 'container': data.container.images[3] must be valid exactly by one definition (0 matches found) Reason of fail for oneOf rule 'cortex_m7_app': data.cortex_m7_app must be file; Non-existing file: /home/jenkins/agent/workspace/zephyr_test_pipe/zephyr/twister-out/imx95_evk_mimx9596_m7/zephyr/tests/misc/test_build/buildsystem.kconfig.utf8_in_values/zephyr/zephyr.bin; The file must exists even if the key is NOT used in configuration. See debug log file: /home/user/.local/state/spsdk/2.6.0/log/debug.log for more info ninja: build stopped: subcommand failed
```

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,47 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
   )
 endif()
+
+if(CONFIG_BOARD_NXP_SPSDK_IMAGE OR (DEFINED ENV{CONFIG_BOARD_NXP_SPSDK_IMAGE}
+  AND "$ENV{CONFIG_BOARD_NXP_SPSDK_IMAGE}" STREQUAL "y"))
```
what's with the ENV parts here?

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,47 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+    ")
+  set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+    COMMAND chmod +x ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin
+    COMMAND ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin --auto-accept --force
```
this needs more limiting, is this x86? x86_64? amd64?

yangbolu1991 replied:
I checked the firmware. It's a self-extractable archive.
It's run in sh bash with bzip2 to extract.
So, I limited it in CMAKE_HOST_UNIX. It's not limited by machine arch.

Thanks.

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -11,3 +11,47 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+            - v2x_dummy: true
+    ")
+  set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+    COMMAND chmod +x ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin
```
and actually it should be added to the repo with this already, do not chmod here

yangbolu1991 replied:
The problem is the firmware which is a self-extractable archive is not from any repo. But from web site.
https://www.nxp.com.cn/lgfiles/NMG/MAD/YOCTO/imx95-19x19-lpddr5-evk-boot-firmware-0.1.bin

So, actually I have no idea. Not sure if there is other way to process it.

nordicjm replied:
@pdgendt @carlescufi @nashif what are your thoughts on this being a linux-only executable that needs a chmod?

pdgendt replied:
I'd be hesitant of running an executable downloaded during the build process. Would it be possible to have a different format for the required files?

I also opened https://github.com/zephyrproject-rtos/zephyr/issues/62250 a long time ago.

JiafeiPan replied:
Currently we still don't support the other archive format, so I think we can merge this PR firstly and provide update if a better archive format is available for us, thanks.

nordicjm commented on boards/nxp/imx95_evk/board.cmake:
```diff
@@ -3,3 +3,14 @@
   board_set_rimage_target(imx95)
 endif()
+
+if(CONFIG_BOARD_NXP_SPSDK_IMAGE OR (DEFINED ENV{CONFIG_BOARD_NXP_SPSDK_IMAGE}
+  AND "$ENV{CONFIG_BOARD_NXP_SPSDK_IMAGE}" STREQUAL "y"))
```
as above, `CONFIG_` is a zephyr Kconfig thing, it has no place in an environmental value, this makes no sense

yangbolu1991 replied:
@hakehuang 
Any suggestion here?
Thanks.

hakehuang replied:
you can rename it. @yangbolu1991

yangbolu1991 commented:
> there is one issue found here
> 
> ```
> /home/jenkins/agent/workspace/zephyr_test_pipe/zephyr/twister-out/imx95_evk_mimx9596_m7/zephyr/tests/misc/test_build/buildsystem.kconfig.utf8_in_values/zephyr/A_kconfig_value_with_a_utf8_char_in_it_Be_.elf for board: imx95_evk Welcome to NXP imx95-19x19-lpddr5-evk-boot-firmware-0.1.bin WARNING: EULA has been auto-accepted; this implies that you agree with it. 
> Unpacking file ............... done SPSDKError: SPSDK: Configuration validation failed: data.containers[1] must be valid exactly by one definition (0 matches found) 
> Reason of fail for oneOf rule 'container': data.container.images[3] must be valid exactly by one definition (0 matches found)
> Reason of fail for oneOf rule 'cortex_m7_app': data.cortex_m7_app must be file; Non-existing file: 
> /home/jenkins/agent/workspace/zephyr_test_pipe/zephyr/twister-out/imx95_evk_mimx9596_m7/zephyr/tests/misc/test_build/buildsystem.kconfig.utf8_in_values/zephyr/zephyr.bin;
>  The file must exists even if the key is NOT used in configuration. 
> See debug log file: /home/user/.local/state/spsdk/2.6.0/log/debug.log for more info ninja: build stopped: subcommand failed
> ```

The error message reported zephyr.bin was not there.
I guess you should not generate bootable firmware for this case.

hakehuang commented:
> The error message reported zephyr.bin was not there.
> I guess you should not generate bootable firmware for this case.

it is related to the buildsystem.kconfig.utf8_in_value.

yangbolu1991 commented:
Updated to v9. Changes include,
- Rebased.
- Converted to use env USE_NXP_SPSDK_IMAGE for spsdk control.

Thanks a lot.

JiafeiPan commented:
@nordicjm would you please revisit this PR to check whether your comments have been addressed, thanks.

hakehuang commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,49 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+    family: mimx9596
+    revision: a1
+    target_memory: standard
+    output:                   ${CMAKE_BINARY_DIR}/zephyr/flash.bin
```
we should use CONFIG_KERNEL_BIN_NAME as in tests/misc/test_build there uses below for output binary

```
    extra_configs:
      - CONFIG_KERNEL_BIN_NAME="A_kconfig_value_with_a_utf8_char_in_it_Bøe_"
```
@yangbolu1991

yangbolu1991 commented:
Updated to v10 for review. Changes includes,
- Rebased to address conflict.
- Generated bootable firmware with KERNEL_BIN_NAME file instead of zephyr.bin, as @hakehuang  pointed out.

Thank you very much.

hakehuang commented:
@yangbolu1991 please fix the branch conflicts

pdgendt commented:
@hakehuang why remove the `manifest`, `manifest-hal_nxp` and `DNM (manifest)` tags? The HAL still needs to be updated?

JiafeiPan commented:
rebased and added a-core support.

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,104 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+              - cortex_m7_app:  ${CMAKE_BINARY_DIR}/zephyr/${CONFIG_KERNEL_BIN_NAME}.bin
+              - v2x_dummy: true
+      ")
+    set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+      COMMAND chmod +x ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin
+      COMMAND ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin --auto-accept --force
+      COMMAND nxpimage ahab export -c ${CMAKE_BINARY_DIR}/zephyr/${AHAB_CONFIG_FILE}
+    )
```
This is downloading a file, changing permissions and executing it and is a binary blob. Binary blobs needs tsc approval @nashif and for what is happening here would like @ceolin to have a look

hakehuang commented:
> @hakehuang why remove the `manifest`, `manifest-hal_nxp` and `DNM (manifest)` tags? The HAL still needs to be updated?

sorry, my bad, I did not see the hal pr is not merged yet. @pdgendt

JiafeiPan commented:
It seems twister failure is not caused by this PR, rebased and run twister again.

hakehuang commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,104 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+    set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+      COMMAND chmod +x ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin
+      COMMAND ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin --auto-accept --force
+      COMMAND nxpimage -v bootable-image merge -c ${CMAKE_BINARY_DIR}/zephyr/imx95_evk_mimx9596_a55_ahab_flash_template.yaml -o flash.bin
```
this command line is change to `nxpimage -v bootable-image export` see M7 above

JiafeiPan replied:
Confirmed from SPSDK v3.0.0 it will change from merge command to export command, so added version checking to adapt to different SPSDK version, thanks.

JiafeiPan commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_m7.yaml:
```diff
@@ -19,4 +19,7 @@ supported:
   - netif:eth
+testing:
+  binaries:
+    - flash.bin
```
will update it, thanks.

hakehuang commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,121 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+        - container:
+            srk_set: none
+            images:
+              - lpddr_imem:     ${CMAKE_BINARY_DIR}/zephyr/${FIRMWARE_RELEASE}/lpddr5_imem_v202311.bin
```
FIRMWARE_RELEASE need check availability

hakehuang commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -115,88 +115,139 @@ Programming and Debugging (A55)
+.. code-block:: console
 
-1. Build and run the Non-SMP application
+   west blobs fetch hal_nxp
```
as for now, we need manually unzip the blobs, so need add steps

hakehuang commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,121 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+              - v2x_dummy: true
+      ")
+    set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+      COMMAND chmod +x ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin
```
Should we remove this?

hakehuang commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,121 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+      secondary_image_container_set: ${CMAKE_BINARY_DIR}/zephyr/imx95_evk_mimx9596_a55_ahab_secondary.yaml
+   ")
+    set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+      COMMAND chmod +x ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin
```
same here, we should remove this

yangbolu1991 committed:
scripts: requirements: extras: add spsdk

Secure Provisioning SDK (SPSDK) is a unified, reliable,
and easy to use Python SDK library working across the
NXP MCU portfolio providing a strong foundation from
quick customer prototyping up to production deployment.

Docs: https://spsdk.readthedocs.io

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

yangbolu1991 committed:
scripts: west_commands: runners: add nxp spsdk

Added nxp spsdk as west runner.

Secure Provisioning SDK (SPSDK) is a unified, reliable,
and easy to use Python SDK library working across the
NXP MCU portfolio providing a strong foundation from
quick customer prototyping up to production deployment.

Docs: https://spsdk.readthedocs.io

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

yangbolu1991 committed:
west.yml: update hal_nxp to support imx95 blobs

Updated hal_nxp to support imx95 blobs.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Rebased to address conflict and fix compliance issue.
Thanks.

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,121 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+      COMMAND 7z x imx-boot-firmware/${FIRMWARE_RELEASE} -aos -bso0 -bse1
+      COMMAND nxpimage ahab export -c ${CMAKE_BINARY_DIR}/zephyr/${AHAB_CONFIG_FILE}
+    )
+  elseif (CONFIG_SOC_MIMX9596_A55)
```
no space between if and bracket in cmake

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,121 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
 
+if(CONFIG_BOARD_NXP_SPSDK_IMAGE OR (DEFINED ENV{USE_NXP_SPSDK_IMAGE}
+  AND "$ENV{USE_NXP_SPSDK_IMAGE}" STREQUAL "y"))
+  find_program(7Z_EXECUTABLE 7z)
```
use reequired flag

yangbolu1991 replied:
Will use it. Thanks.

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -12,5 +12,121 @@ if (CONFIG_SOF AND CONFIG_BOARD_IMX95_EVK_MIMX9596_M7_DDR)
+              - v2x_dummy: true
+      ")
+    set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+      COMMAND 7z x ${ZEPHYR_HAL_NXP_MODULE_DIR}/zephyr/blobs/imx-boot-firmware/${FIRMWARE_RELEASE}.bin -o./imx-boot-firmware -aos -bso0 -bse1
```
use the variable returned by find_program() for all invocations

yangbolu1991 replied:
Will use it. Thanks.

carlescufi reviewed (changes requested):
Please move [scripts: blobs: add license-click-through property](https://github.com/zephyrproject-rtos/zephyr/pull/80507/commits/e10aa0b0a25e8d458090563c7dbd879e34749f17) to its own PR. We can get it merged quickly, but I would rather have it separate from the rest.

yangbolu1991 commented:
> Please move [scripts: blobs: add license-click-through property](https://github.com/zephyrproject-rtos/zephyr/pull/80507/commits/e10aa0b0a25e8d458090563c7dbd879e34749f17) to its own PR. We can get it merged quickly, but I would rather have it separate from the rest.

Thanks. I create a separate PR for this.
https://github.com/zephyrproject-rtos/zephyr/pull/90136

yangbolu1991 commented:
Updated PR to new version. Changes include,
- Rebased to address conflict.
- Code improvement per @nordicjm  comments.

Thanks.

nordicjm commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -115,88 +115,139 @@ Programming and Debugging (A55)
+
+   It is recommended running the command above after :file:`west update`.
+
+SPSDK runner is enabled by configure item ``CONFIG_BOARD_NXP_SPSDK_IMAGE``, currently
```
use proper syntax for all Kconfig items i.e.
```
:kconfig:option:`CONFIG_BOARD_NXP_SPSDK_IMAGE`
```

yangbolu1991 replied:
Thanks @nordicjm 
Updated.

JiafeiPan commented:
> Please move [scripts: blobs: add license-click-through property](https://github.com/zephyrproject-rtos/zephyr/pull/80507/commits/e10aa0b0a25e8d458090563c7dbd879e34749f17) to its own PR. We can get it merged quickly, but I would rather have it separate from the rest.

Hi, @carlescufi , would you please check your changes requested has been addressed in this PR? thanks.

dleach02 commented:
Please rebase the HAL PR

JiafeiPan commented:
> Please rebase the HAL PR

Rebased.

JiafeiPan commented:
Rebased this PR to trigger twister testing.

yangbolu1991 commented:
Rebased to address west.yml conflict.
Thanks.

hakehuang commented:
> Rebased to address west.yml conflict. Thanks.

@yangbolu1991 M7 seems not working, please check. I test with hello_world, it does not starts well. but A core is OK.

carlescufi reviewed (approved):
LGTM, thanks
